File: trees/max_depth_of_bt.py
from typing import Optional
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    # dfs
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        l = 1+self.maxDepth(root.left)
        r = 1+self.maxDepth(root.right)
        return max(l,r)

    #bfs
    def maxDepth2(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        res = 0
        q = deque([root])
        while q:
            res += 1
            for _ in range(len(q)):
                node = q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        return res
        

File: trees/max_depth_of_bt.java
package trees;

public class max_depth_of_bt {
    
}

//Definition for a binary tree node.
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    // dfs
    public int maxDepth(TreeNode root) {
        if(root == null){
            return 0;
        }
        int l = 1 + maxDepth(root.left);
        int r = 1 + maxDepth(root.right);
        return Math.max(l, r);
    }
    
    // bfs
    public int maxDepth2(TreeNode root){
        if(root == null){
            return 0;
        }
        Deque<TreeNode> dq = new ArrayDeque<>();
        dq.offer(root);
        int res = 0;

        while(!dq.isEmpty()){
            res++;
            int size = dq.size();
            for(int i = 0; i < size; i++){
                TreeNode node = dq.poll();
                if(node.left != null){
                    dq.offer(node.left);
                }
                if(node.right != null){
                    dq.offer(node.right);
                }
            }
        }

        return res;
    }
}